specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json

```diff
@@ -0,0 +1,379 @@
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "Extension release train: preview or stable"
+        },
+        "clusterTypes": {
+          "description": "Cluster types",
+          "readOnly": true,
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "supportedScopes": {
+          "readOnly": true,
+          "type": "object",
+          "description": "Extension scopes",
+          "$ref": "#/definitions/SupportedScopes"
+        }
+      }
+    },
+    "ExtensionVersionList": {
+      "description": "List versions for an Extension",
+      "type": "object",
+      "properties": {
+        "value": {
+          "description": "Versions available for this Extension Type",
+          "type": "array",
+          "items": {
+            "type": "object",
+            "properties": {
+              "releaseTrain": {
+                "description": "The release train for this Extension Type",
+                "type": "string"
+              },
+              "versions": {
+                "type": "array",
+                "description": "Versions available for this Extension Type and release train",
+                "items": {
+                  "type": "string"
+                }
+              }
+            }
+          }
+        },
+        "nextLink": {
+          "description": "The link to fetch the next page of Extension Types",
+          "type": "string"
+        },
+        "systemData": {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
+          "readOnly": true,
+          "description": "Metadata pertaining to creation and last modification of the resource."
+        }
+      }
+    },
+    "SupportedScopes": {
+      "type": "object",
+      "description": "Extension scopes",
+      "properties": {
+        "defaultScope": {
+          "type": "string",
+          "description": "Default extension scopes: cluster or namespace"
+        },
+        "clusterScopeSettings": {
+          "description": "Scope settings",
+          "$ref": "#/definitions/ClusterScopeSettings"
+        },
+        "namespaceScopeSettings": {
+          "description": "Scope settings",
+          "$ref": "#/definitions/NamespaceScopeSettings"
+        }
+      }
+    },
+    "ClusterScopeSettings": {
+      "type": "object",
+      "description": "Extension scope settings",
+      "properties": {
+        "allowMultipleInstances": {
+          "type": "boolean",
+          "description": "Describes if multiple instances of the extension are allowed"
+        },
+        "defaultReleaseNamespace": {
+          "type": "string",
+          "description": "Default extension release namespace"
+        }
+      }
+    },
+    "NamespaceScopeSettings": {
+      "type": "object",
+      "description": "Extension scope settings",
+      "properties": {}
```

Comment:
>properties": {} [](http://example.com/codeflow?start=7&length=15)

ARM recommends against the use of the additionalProperties/empty/unschematized object pattern. ARM favors stronger typing wherever possible because it allows all consumers of your API to reason about it. This has benefits for everyone that interacts with your API including basic infrastructure, other Azure features, SDK owners, and customers.

For example, customers will not be able to manage these objects using Azure Policy, which is a standard ARM feature. Your customers may be unhappy about that. There are other features like ARG and others under development that won't be able to support these objects. Unschematized objects are essentially API contracts that bypass Azure versioning infrastructure.

Generally, this pattern should only be used when needed and the object properties are dynamic, unknown, or user defined. To address this you can either enumerate the object's properties in your swagger, implement it as its own resource type, or remove the object from the spec.

Ref: [https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined).

Reply:
Good point I will remove this property.